WIP: Fix improper error message in Export-CSV

Maamue opened the pull request:
# PR Summary

- Rework the ParameterSets in Export-CSV to reflect behaviour from
Import-CSv.
- Make -Path and -LiteralPath manadatory in their sets.
- Remove error message from Export-CSV when both -Path and -LiteralPath
are specified, because they can no lnger be specified at the same time.

fixes #11679

## PR Context
Rework ParameterSets to behave like Import-CSV. Path and LiteralPath can not be specified at the same time. ParamaterSets themselves can validate if a proper set was given or not. No further checks in the code are necessary. 

No issue filed for documentation, but prepared the necessary PR https://github.com/MicrosoftDocs/PowerShell-Docs/pull/5948

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

Maamue committed:
Fix improper error message in Export-CSV

- Rework the ParameterSets in Export-CSV to reflect behaviour from
Import-CSv.
- Make -Path and -LiteralPath manadatory in their sets.
- Remove error message from Export-CSV when both -Path and -LiteralPath
are specified, because they can no lnger be specified at the same time.

Maamue committed:
Fix improper error message in Export-CSV

- Rework the ParameterSets in Export-CSV to reflect behaviour from
Import-CSv.
- Make -Path and -LiteralPath manadatory in their sets.
- Remove error message from Export-CSV when both -Path and -LiteralPath
are specified, because they can no lnger be specified at the same time.

Maamue committed:
Fix An opening curly bracket must not be followed by a blank line.

Maamue commented:
Tests in test\powershell\Modules\Microsoft.PowerShell.Utility\ImportExportCSV.Delimiter.Tests.ps1 fail. Will add fixes to the PR.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CsvCommands.cs:
```diff
@@ -244,14 +246,6 @@ protected override void BeginProcessing()
-            // Validate that they don't provide both Path and LiteralPath, but have provided at least one.
-            if (!(_specifiedPath ^ _isLiteralPath))
-            {
-                InvalidOperationException exception = new InvalidOperationException(CsvCommandStrings.CannotSpecifyPathAndLiteralPath);
```
You should remove the resource string from resx file too.

vexx32 replied:
@Maamue if you have trouble finding that, do a search through the repo for the string name (`CannotSpecifyPathAndLiteralPath`) and you're looking for the resx file as Ilya mentioned, not the C# one (which is generated from the resx file at build time). 🙂

Maamue replied:
Thank you for the hints! I found it and removed it. Tests are in progress :)

Maamue committed:
CSVCommand: Remove unused Resource string.

Remove the unused resource string for the validation, whether Path and
LiteralPath were used at the same time. With the changed ParameterSets
this is no longer necessary

Maamue committed:
CsvCommands.cs: Remove positional argument for parameter LiteralPath

- Remove positional argument for parameter Literalpath. Otherwise it
can not resolve the ParameterSet for positional binding when parameter
UseCulture is used but the parameter (Literal)Path is posiitonal.
- Add mandatory argument to parameter UseCulture

Maamue committed:
Export-Csv.Tests.ps1: Split LiteralPath parameter tests in two

With the new ParameterSets this test no longer works since it can not
be resolved. It must be called specifically with either Delimiter or
UseCulture. The Delimiter chosen is arbitrary, since the Delimiter
tests happen in ImportExportCSVDelimiter.Tests.ps1.

Remove test for ErrorId "CannotSpecifyPathAndLiteralPath". This error
will not be thrown by the ParameterBinding.

Maamue commented:
I thought with a simple ParameterSet for Path and LiteralPath this would be solved. But this is not enough. With the additional ParameterSets the Cmdlet does not know how it it should behave in certain conditions (e.g. when only provided with LiteralPath and InputObject from the Pipeline). 

Basically the in-cmdlet checks that are there for Path/LiteralPath need to be reworked to apply to to Delimiter/UseCulture. So we only shift the symptom to other Parameters. I have not yet found a way that it works with every test passing. I'm not sure I will. If nobody wants to chime in I suggest closing this PR and when I / someone else figured out how to deal with this we can either reopen it or make a new PR. 

While it may not be a successful endeavour, I at least learned how the repo and tooling here works :)

vexx32 commented:
@Maamue sounds like we might need to rewrite the parameter binder first, eh? 😄 

Perhaps something will come of #12818. We'll see, I suppose. ^^

Maamue commented:
Closing
